hs/src/Reach/Eval/Core.hs

```diff
@@ -4124,14 +4128,45 @@ evalPrim p sargs =
     SLPrim_Contract_new -> do
       at <- withAt id
       let lab = "new Contract"
-      (ccv, opts) <-
+      let expectContractCode = \case
+            SLV_ContractCode _ cc -> return cc
+            ow -> expect_t ow $ Err_Expected "ContractCode or Reach.App"
+      (cc, opts) <-
         case args of
-          [x] -> return (x, mempty)
-          [x, y] -> (,) x <$> mustBeObject y
+          [x@(SLV_Prim (SLPrim_App_Delay _ _ _ _ mname))] -> do
+            -- Compile Reach.App and remember result of compilation
+            compileProg <- asks e_compileProg
+            name <- maybe (("Reach.App" <>) . show <$> ctxt_alloc) return mname
+            co <- liftIO $ compileProg name x
+
+            -- Tweak the output of compilation to generate objects with the keys we need
+            let get k = fromMaybe (impossible $ "No key: " <> show k) . KM.lookup k
+            let slssv = SLSSVal at Public
+            let getUInt k v = do
+                  case get k v of
+                    AE.Number n -> slssv $ SLV_Int at Nothing $ coefficient n
+                    _ -> impossible "Expected UInt"
+            let goAlgo = expectJsonObj $ \ o ->
+                  Object $ KM.union o $
+                            KM.fromList [ ("approvalB64", get "appApproval" o)
+                                        , ("clearStateB64", get "appClear" o) ]
+            let goAlgoOpts :: Value -> SLEnv
+                goAlgoOpts = expectJsonObj $ \ o ->
+                            M.fromList [ ("globalUints", getUInt "GlobalNumUint" o)
+                                        , ("globalBytes", getUInt "GlobalNumByteSlice" o)
+                                        , ("localBytes", getUInt "LocalNumByteSlice" o)
+                                        , ("localUints", getUInt "LocalNumUint" o) ]
+            let goEth = expectJsonObj $ mapJsonString (T.drop 2) . get "Bytecode"
+            let co' = flip M.mapWithKey co $ \case
+                        "ALGO" -> goAlgo
+                        _  -> goEth
+            let opts'= M.fromList $ flip map (M.toList co) $ \ (k, v) -> case k of
```

Comment:
This is very scandalous and should at least be in ALGO and part of the connector interface like conGen